### ISSUE TITLE
How to render GitLab's  "!reference [some, rule]" line with ytt

### ISSUE THREAD

**Issue by astaruch** [k14s/ytt#688]
Hello,

I am struggling with the ytt and advanced use of GitLab's  [child-parent pipelines](https://docs.gitlab.com/ee/ci/pipelines/parent_child_pipelines.html#dynamic-child-pipelines).

I want to use [`!reference`](https://docs.gitlab.com/ee/ci/yaml/yaml_optimization.html#reference-tags) tag feature in the child pipeline, that will be generated by the `ytt`. 

Let's go into the playground - https://carvel.dev/ytt/#example:example-plain-yaml:

When I define the following simple YAML file:
```yaml
rules:
  some_rule:
    - if: "$CI_COMMIT_TAG"

some_node: 
  - !reference [rules, some_rule]
```

The `ytt` ignores the `!reference` and renders the expression behind it as an array:
```yaml
rules:
  some_rule:
  - if: $CI_COMMIT_TAG
some_node:
- - rules
  - some_rule
```

Now, how can I achieve that `ytt` will render the following yaml file:
```
rules:
  some_rule:
  - if: $CI_COMMIT_TAG
some_node:
- !reference [rules, some_rule]
```
?

I have tried following but all yields unsatisfactory results:
```
- !!reference [rules, some_rule]   # no change
- \!reference [rules, some_rule]   # - \!reference [rules, some_rule] -> invalid syntax for GitLab CI
- #@ "!reference [rules, some_rule]"  #  - '!reference [rules, some_rule]'  -> not wanted - it is a string and not a !reference node
- "!reference [rules, some_rule]"        #   - '!reference [rules, some_rule]'  -> not wanted - it is a string and not a !reference node
```

Is even what I want achievable with `ytt`?

This simple example doesn't contain anything to the `ytt`, since it would be irrelevant to this example, but I can post my it if needed.

Thanks for the help.

Regards,
Andrej

**Comment by astaruch** [comment]
After reading the [YAML specification for tags](https://yaml.org/spec/1.2.2/#tags), I understand that:
- `!reference` is denoting a special tag created by GitLab (https://gitlab.com/gitlab-org/gitlab/-/issues/266173)
- since `ytt` is rendering it and doesn't know anything about types, it is just ignoring it and displaying it as an array

I am thinking about the two following workarounds directly with `ytt`:

1. Turn off `ytt` for a single line, e.g.
```
some_node:
- !reference [rules, some_rule] #@ some_command_to_say_keep_this_line_unattached
```

2. Have a function, that can return a string without quotation marks:
```
some_node:
- #@ return_str_without_quots("!reference [rules, some_rule]")
```

Or is there another way how to achieve my goal with `ytt`?

I am now defining in the `yaml` string, and then remove single quotes with `sed`, but that seems to me like a hacky solution:
```
cat <<EOF > yaml.yml
rules:
  some_rule:
    - if: "$CI_COMMIT_TAG"

some_node: 
  - "!reference [rules, some_rule]"
EOF

# Incorrect Gitlab CI file
$ ytt -f yaml.yml 
rules:
  some_rule:
  - if: ""
some_node:
- '!reference [rules, some_rule]'

# Correct GitLab CI file
$ ytt -f yaml.yml | sed -e "s/'\(\!reference.*\)'/\1/"
rules:
  some_rule:
  - if: ""
some_node:
- !reference [rules, some_rule]
```

**Comment by pivotaljohn** [comment]
Hello @astaruch.

TL;DR. `ytt` does not (currently) emit tags in its output.

Yeah, that there is a YAML tag... as you noted, it's parsed and processed by `ytt` as such. This is because `ytt` operates _from within_ the YAML structure. In fact, first step — before _any_ processing occurs — `ytt` delegates parsing the YAML to the de facto standard Go YAML parser.
On the output side, `ytt` currently does _not_ include tag information when handing off to the same YAML library to marshal the result.

...

In case you're not tied to the `!reference` mechanism, I believe you can achieve the same re-use if the referenced/referencing fragments were part of your `ytt` templating...

From https://gitlab.com/gitlab-org/gitlab/-/issues/266173:
```yaml
# Content of setup.yml
.setup:
  script:
    - echo creating environment1 
    - echo creating environment2

# Content of teardown.yml
.teardown:
  script:
    - echo deleting environment1
    - echo deleting environment2

# .gitlab-ci.yml
include:
    local: setup.yml
    local: teardown.yml
 
test:
  script:
    - !reference [.setup, script] # or any other name that would allow us to "include in this context some other value"
    - echo running my own command
    - !reference [.teardown, script]
```

You could extract those fragments and included them in `ytt` as a `.lib.yaml` file:

```yaml
#! Content of helpers.lib.yaml
---
#@ def setup():
- echo creating environment1 
- echo creating environment2
#@ end
---
#@ def teardown():
- echo deleting environment1
- echo deleting environment2
#@ end

#! .gitlab-ci.yml
#@ load("@ytt:template", "template")
#@ load("helpers.lib.yaml", "setup", "teardown")

test:
  script:
    - #@ template.replace(setup())
    - echo running my own command
    - #@ template.replace(teardown())
```
_([in ytt playground](https://carvel.dev/ytt/#gist:https://gist.github.com/pivotaljohn/a7ac9e295dbebe00b0aa15a76f7ce42a))_

Any reason that might not work for your situation?

**Comment by pivotaljohn** [comment]
It's been a minute. Closing for now. @astaruch, if you'd still like support with this issue, feel free to comment on it, we'll get notified. 👍🏻

**Comment by sb3lxb** [comment]
I ended up using `sed` too.  To simplify the process and tell YTT not to quote the value, write it as `\! reference [...]`.  Then the cleanup is easier to read--`sed -i 's/\\!reference/!reference/g'  FILE_NAME`.

Update: correction.  The quotes will be required if you use the YTT `--strict` option.
